File: sphone-api/Domain/Orders/InstallmentPlan.cs
using SPhone.Domain.Common;

namespace SPhone.Domain.Orders;

public sealed class InstallmentPlan : ValueObject
{
    public decimal ProductPrice { get; }
    public decimal DownPayment { get; }
    public int TotalMonths { get; }
    public decimal InterestRate { get; } // annual interest rate as decimal e.g. 0.18 = 18%
    public decimal MonthlyAmount { get; }
    public decimal TotalAmount { get; }

    public InstallmentPlan(decimal productPrice, decimal downPayment, int totalMonths, decimal interestRate = 0.18m)
    {
        if (productPrice <= 0) throw new DomainException("Product price must be positive.");
        if (downPayment < 0) throw new DomainException("Down payment cannot be negative.");
        if (downPayment >= productPrice) throw new DomainException("Down payment cannot exceed product price.");
        if (totalMonths is not (3 or 6 or 12 or 18 or 24))
            throw new DomainException("Installment plan must be 3, 6, 12, 18, or 24 months.");

        ProductPrice = productPrice;
        DownPayment = downPayment;
        TotalMonths = totalMonths;
        InterestRate = interestRate;

        var principal = productPrice - downPayment;
        var monthlyRate = interestRate / 12;
        MonthlyAmount = monthlyRate == 0
            ? principal / totalMonths
            : Math.Ceiling(principal * monthlyRate * (decimal)Math.Pow((double)(1 + monthlyRate), totalMonths)
                / ((decimal)Math.Pow((double)(1 + monthlyRate), totalMonths) - 1));

        TotalAmount = downPayment + MonthlyAmount * totalMonths;
    }

    protected override IEnumerable<object?> GetEqualityComponents()
    {
        yield return ProductPrice;
        yield return DownPayment;
        yield return TotalMonths;
        yield return InterestRate;
    }
}

File: sphone-api/src/domain/orders/InstallmentPlan.ts
const ALLOWED_TERMS = [3, 6, 12, 18, 24]
const ANNUAL_INTEREST_RATE = 0.18

export class InstallmentPlan {
  readonly productPrice: number
  readonly downPayment: number
  readonly totalMonths: number
  readonly interestRate: number
  readonly monthlyAmount: number
  readonly totalAmount: number

  private constructor(params: {
    productPrice: number
    downPayment: number
    totalMonths: number
    interestRate: number
    monthlyAmount: number
    totalAmount: number
  }) {
    this.productPrice = params.productPrice
    this.downPayment = params.downPayment
    this.totalMonths = params.totalMonths
    this.interestRate = params.interestRate
    this.monthlyAmount = params.monthlyAmount
    this.totalAmount = params.totalAmount
  }

  static calculate(productPrice: number, downPayment: number, totalMonths: number): InstallmentPlan {
    if (!ALLOWED_TERMS.includes(totalMonths)) {
      throw new Error(`Total months must be one of: ${ALLOWED_TERMS.join(', ')}`)
    }
    if (downPayment <= 0 || downPayment >= productPrice) {
      throw new Error('Down payment must be greater than 0 and less than product price')
    }

    const principal = productPrice - downPayment
    const monthlyRate = ANNUAL_INTEREST_RATE / 12
    // PMT formula: P * r * (1+r)^n / ((1+r)^n - 1)
    const factor = Math.pow(1 + monthlyRate, totalMonths)
    const monthlyAmount = Math.round((principal * monthlyRate * factor) / (factor - 1) * 100) / 100
    const totalAmount = Math.round(monthlyAmount * totalMonths * 100) / 100

    if (monthlyAmount < 500) {
      throw new Error('Monthly installment must be at least 500 THB')
    }

    return new InstallmentPlan({
      productPrice,
      downPayment,
      totalMonths,
      interestRate: ANNUAL_INTEREST_RATE,
      monthlyAmount,
      totalAmount,
    })
  }

  static reconstitute(params: {
    productPrice: number
    downPayment: number
    totalMonths: number
    interestRate: number
    monthlyAmount: number
    totalAmount: number
  }): InstallmentPlan {
    return new InstallmentPlan(params)
  }
}
